galerautils/src/gu_enc_pmemory_manager.cpp

```diff
@@ -0,0 +1,175 @@
+    static const std::size_t nbytes = sysconf(_SC_PAGESIZE);
+    return nbytes;
+};
+
+PMemoryManager::PMemoryManager(size_t size, size_t allocPageSize)
+: createSize_(size)  // actual size may differ because of limits
+, base_(0)
+, size_(0)
+, freePages_()
+, myPages_()
+, fd_(-1)
+, mapped_(false)
+, allocPagesCnt_(0)
+, allocPageSize_(allocPageSize) {
+    S_DEBUG_A("+++PMemoryManager::PMemoryManager() size: %ld, allocPageSize: %ld\n",
+      size, allocPageSize);
+
+    // allocPageSize has to be CPU page aligned
+    if (allocPageSize_ % getCpuPageSize() || allocPageSize_ < getCpuPageSize()) {
+        S_DEBUG_A("PMemoryManager::PMemoryManager() allocPageSize not aligned. Requested: %ld. "
+                  "Should be multiply of CPU page size %ld\n", allocPageSize, getCpuPageSize());
+        gu_throw_error(errno) << "PMemoryManager::PMemoryManager() allocPageSize not aligned";
+    }
+
+    // how many pages do we need to satisfy size?
+    allocPagesCnt_ = size / allocPageSize_;
+    if (size % allocPageSize_) {
+        S_DEBUG_A("PMemoryManager::PMemoryManager() adding page, size %ld is not aligned to allocation unit\n", size);
+        allocPagesCnt_++;
+    }
+    allocPagesCnt_ = allocPagesCnt_ < CACHE_ALLOC_PAGES_MAX ? allocPagesCnt_ : CACHE_ALLOC_PAGES_MAX;
+
+    size_ = allocPagesCnt_ * allocPageSize_;
+    if (createTmpFile()) {
+        gu_throw_error(errno) << "PMemoryManager::PMemoryManager() creation of tempfile failed";
+    }
+    base_ = static_cast<unsigned char*>(mmap(nullptr, size_, PROT_READ|PROT_WRITE, MAP_SHARED, fd_, 0));
+    mapped_ = (base_ != MAP_FAILED);
+    if (!mapped_)
+    {
+        gu_throw_error(errno) << "PMemoryManager::PMemoryManager() mmap() failed";
+    }
+    if (mlock(base_, size_)) {
+        S_DEBUG("PMemoryManager::PMemoryManager() mlock failed. It will still work, "
+                   "but swap pages into the disk, so performance will be affected\n");
+    }
+#if CLEAR_BUFFERS
+    memset(base_, FREE_PAGE_PATTERN, size_);
+#endif
+    S_DEBUG_A("PMemoryManager::PMemoryManager() (x%llX - x%llX). "
+              "CpuPageSize: %ld, allocPageSize: %ld, allocPagesCnt: %ld, "
+              "size requested: %ld, size allocated: %ld\n",
+      ptr2ull(base_), ptr2ull(base_) + size_,
+      getCpuPageSize(), allocPageSize_, allocPagesCnt_, createSize_, size_);
+
+    for (size_t i = 0; i < allocPagesCnt_; ++i) {
+        auto page = std::make_shared<PPage>();
+        page->fd_ = fd_;
+        page->offset_ = i*allocPageSize_;
+        page->ptr_ = base_ + page->offset_;
+        myPages_.push_back(page);
+    }
+    freePages_ = myPages_;
+    S_DEBUG("---PMemoryManager::PMemoryManager()\n");
+}
+
+PMemoryManager::~PMemoryManager() {
+    S_DEBUG_A("+++PMemoryManager::~PMemoryManager() (x%llX - x%llX)\n",
+      ptr2ull(base_), ptr2ull(base_) + size_);
+
+    if (freePages_.size() != allocPagesCnt_) {
+        S_DEBUG_A("Some pages still allocated. Free pages cnt: %d\n", freePages_.size());
```

Comment:
S_DEBUG macros are intended to be used from the signal handler context. We can't use fprintf/existing server logging system, because we need to be lock-free. So unistd.h write() function needs to be used.
By "higher level" do you mean formatting the message like server's [ERROR]/[WARNING], to be automatically catchable by MTR framework?

Reply:
Yes, I meant the error/warning formatting.

Reply:
Implemented.